`Bug` [Samsung_HVAC]: ESPHome 2026.3.0 bug

### The problem

New version of esphome with compilation bug.

### ESPHome Version

2026.3.0

### Samsung HVAC Component Version

main

### Home Assistant Version

2026.3.2

### AC Type

NASA

### Indoor AC Unit Model

AR18CVFAAWKNAZ

### Outdoor AC Unit Model

AR18CVFAAWKXAZ

### ESP Device Model

M5STACK ATOM Lite

### Connection Points

F1/F2

### ESPHome Logs

INFO ESPHome 2026.3.0
INFO Reading configuration /config/esphome/ac-samsung-gustavo.yaml...
INFO Updating https://github.com/omerfaruk-aran/esphome_samsung_hvac_bus.git@main
INFO Generating C++ source...
INFO Setting CONFIG_LWIP_MAX_SOCKETS to 17 (TCP=11 [api=3, captive_portal=3, web_server=5], UDP=3 [captive_portal=1, mdns=2], TCP_LISTEN=3 [api=1, ota=1, web_server_base=1])
INFO Deleting /data/build/samsung-ac-gustavo/.pioenvs/samsung-ac-gustavo/CMakeCache.txt
INFO Compiling app... Build path: /data/build/samsung-ac-gustavo
Processing samsung-ac-gustavo (board: m5stack-atom; framework: espidf; platform: https://github.com/pioarduino/platform-espressif32/releases/download/55.03.37/platform-espressif32.zip)
--------------------------------------------------------------------------------
Created pioarduino Python virtual environment using uv: /root/.platformio/penv
Installed esptool from tl-install path: /data/cache/platformio/packages/tool-esptoolpy
Tool Manager: Installing https://github.com/pioarduino/esp-idf/releases/download/v5.5.3.1/esp-idf-v5.5.3.1.tar.xz
INFO Installing https://github.com/pioarduino/esp-idf/releases/download/v5.5.3.1/esp-idf-v5.5.3.1.tar.xz
Downloading  [####################################]  100%          
Unpacking  [####################################]  100%          
Tool Manager: framework-espidf@3.50503.0 has been installed!
INFO framework-espidf@3.50503.0 has been installed!
Removing unused dependencies...
Library Manager: Removing noise-c @ 0.1.10
INFO Removing noise-c @ 0.1.10
Library Manager: Removing dependencies...
INFO Removing dependencies...
Library Manager: Removing libsodium @ 1.10020.7
INFO Removing libsodium @ 1.10020.7
Library Manager: libsodium@1.10020.7 has been removed!
INFO libsodium@1.10020.7 has been removed!
Library Manager: noise-c@0.1.10 has been removed!
INFO noise-c@0.1.10 has been removed!
Library Manager: Installing esphome/noise-c @ 0.1.11
INFO Installing esphome/noise-c @ 0.1.11
Downloading  [####################################]  100%
Unpacking  [####################################]  100%
Library Manager: noise-c@0.1.11 has been installed!
INFO noise-c@0.1.11 has been installed!
Library Manager: Resolving dependencies...
INFO Resolving dependencies...
Library Manager: Installing esphome/libsodium @ 1.10021.0
INFO Installing esphome/libsodium @ 1.10021.0
Downloading  [####################################]  100%
Unpacking  [####################################]  100%
Library Manager: libsodium@1.10021.0 has been installed!
INFO libsodium@1.10021.0 has been installed!
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - contrib-piohome @ 3.4.4 
 - framework-espidf @ 3.50503.0 (5.5.3) 
 - tool-cmake @ 4.0.3 
 - tool-esp-rom-elfs @ 2024.10.11 
 - tool-esptoolpy @ 5.1.2 
 - tool-ninja @ 1.13.1 
 - tool-scons @ 4.40801.0 (4.8.1) 
 - toolchain-xtensa-esp-elf @ 14.2.0+20251107
Creating a new virtual environment for IDF Python dependencies using uv
Using Python 3.12.10 environment at: /root/.platformio/penv/.espidf-5.5.3
Installing ESP-IDF's Python dependencies with uv
Using Python 3.12.10 environment at: /root/.platformio/penv/.espidf-5.5.3
Resolved 27 packages in 627ms
Downloading cryptography (4.0MiB)
Downloading pydantic-core (2.0MiB)
 Downloaded pydantic-core
 Downloaded cryptography
Prepared 27 packages in 551ms
Installed 27 packages in 86ms
 + annotated-types==0.7.0
 + certifi==2026.2.25
 + cffi==2.0.0
 + charset-normalizer==3.4.6
 + click==8.3.1
 + colorama==0.4.6
 + cryptography==44.0.3
 + esp-idf-kconfig==2.5.3
 + idf-component-manager==2.4.8
 + idna==3.11
 + jsonref==1.1.0
 + psutil==7.2.2
 + pycparser==3.0
 + pydantic==2.12.5
 + pydantic-core==2.41.5
 + pydantic-settings==2.13.1
 + pyparsing==3.3.2
 + python-dotenv==1.2.2
 + requests==2.32.5
 + requests-file==3.0.1
 + requests-toolbelt==1.0.0
 + ruamel-yaml==0.19.1
 + tqdm==4.67.3
 + truststore==0.10.4
 + typing-extensions==4.15.0
 + typing-inspection==0.4.2
 + urllib3==1.26.20
Reading CMake configuration...
Generating assembly for certificate bundle...
Dependency Graph
|-- noise-c @ 0.1.11
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/api/api_buffer.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/api/api_connection.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/api/api_frame_helper.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/api/api_frame_helper_noise.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/api/api_pb2.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/api/api_pb2_service.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/api/api_server.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/api/list_entities.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/api/proto.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/api/subscribe_state.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/captive_portal/captive_portal.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/captive_portal/dns_server_esp32_idf.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/climate/climate.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/climate/climate_mode.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/climate/climate_traits.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/esp32/core.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/esp32/crash_handler.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/esp32/gpio.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/esp32/helpers.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/esp32/preferences.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/esp32/printf_stubs.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/esp32/throw_stubs.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/esphome/ota/ota_esphome.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/json/json_util.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/logger/logger.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/logger/logger_esp32.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/logger/task_log_buffer_esp32.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/logger/task_log_buffer_zephyr.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/md5/md5.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/mdns/mdns_component.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/mdns/mdns_esp32.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/network/util.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/number/automation.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/number/number.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/number/number_call.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/number/number_traits.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/ota/ota_backend.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/ota/ota_backend_esp_idf.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/ota/ota_backend_host.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/safe_mode/safe_mode.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/samsung_ac/conversions.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/samsung_ac/debug_mqtt.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/samsung_ac/log.cpp.o
Compiling .pioenvs/samsung-ac-gustavo/src/esphome/components/samsung_ac/protocol.cpp.o
src/esphome/components/samsung_ac/protocol.cpp: In function 'uint16_t esphome::samsung_ac::skip_data(std::vector<unsigned char>&, int)':
src/esphome/components/samsung_ac/protocol.cpp:39:29: error: no matching function for call to 'find(__gnu_cxx::__normal_iterator<unsigned char*, std::vector<unsigned char> >, std::vector<unsigned char>::iterator, int)'
   39 |             return std::find(data.begin() + from, data.end(), 0x32) - data.begin();
      |                    ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/locale_facets.h:48,
                 from /data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_ios.h:37,
                 from /data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/ios:46,
                 from /data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/ostream:40,
                 from /data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/iostream:41,
                 from src/esphome/components/samsung_ac/util.h:4,
                 from src/esphome/components/samsung_ac/protocol.h:5,
                 from src/esphome/components/samsung_ac/protocol.cpp:1:
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/streambuf_iterator.h:435:5: note: candidate: 'template<class _CharT2> typename __gnu_cxx::__enable_if<std::__is_char<_CharT2>::__value, std::istreambuf_iterator<_CharT, std::char_traits<_CharT> > >::__type std::find(istreambuf_iterator<_CharT, char_traits<_CharT> >, istreambuf_iterator<_CharT, char_traits<_CharT> >, const _CharT2&)'
  435 |     find(istreambuf_iterator<_CharT> __first,
      |     ^~~~
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/streambuf_iterator.h:435:5: note:   template argument deduction/substitution failed:
src/esphome/components/samsung_ac/protocol.cpp:39:29: note:   '__gnu_cxx::__normal_iterator<unsigned char*, std::vector<unsigned char> >' is not derived from 'std::istreambuf_iterator<_CharT, std::char_traits<_CharT> >'
   39 |             return std::find(data.begin() + from, data.end(), 0x32) - data.begin();
      |                    ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*** [.pioenvs/samsung-ac-gustavo/src/esphome/components/samsung_ac/protocol.cpp.o] Error 1
========================= [FAILED] Took 158.59 seconds =========================

### Samsung AC Logs

_No response_

### YAML Configuration

```yaml

```

### Additional Details

_No response_

I have the same issue, same errors

INFO ESPHome 2026.3.0
INFO Reading configuration /config/esphome/tepelko.yaml...
INFO Updating https://github.com/omerfaruk-aran/esphome_samsung_hvac_bus.git@main
INFO Generating C++ source...
INFO Setting CONFIG_LWIP_MAX_SOCKETS to 17 (TCP=11 [api=3, captive_portal=3, web_server=5], UDP=3 [captive_portal=1, mdns=2], TCP_LISTEN=3 [api=1, ota=1, web_server_base=1])
INFO Deleting /data/build/tepelko/.pioenvs/tepelko/CMakeCache.txt
INFO Compiling app... Build path: /data/build/tepelko
Processing tepelko (board: m5stack-atom; framework: arduino, espidf; platform: https://github.com/pioarduino/platform-espressif32/releases/download/55.03.37/platform-espressif32.zip)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - contrib-piohome @ 3.4.4 
 - framework-arduinoespressif32 @ 3.3.7 
 - framework-arduinoespressif32-libs @ 5.5.3 
 - framework-espidf @ 3.50503.0 (5.5.3) 
 - tool-cmake @ 4.0.3 
 - tool-esp-rom-elfs @ 2024.10.11 
 - tool-esptoolpy @ 5.1.2 
 - tool-ninja @ 1.13.1 
 - tool-scons @ 4.40801.0 (4.8.1) 
 - toolchain-xtensa-esp-elf @ 14.2.0+20251107
Reading CMake configuration...
Dependency Graph
|-- noise-c @ 0.1.11
Compiling .pioenvs/tepelko/src/esphome/components/samsung_ac/protocol.cpp.o
Compiling .pioenvs/tepelko/src/esphome/components/samsung_ac/util.cpp.o
Compiling .pioenvs/tepelko/src/esphome/components/select/select.cpp.o
Compiling .pioenvs/tepelko/src/esphome/components/select/select_call.cpp.o
src/esphome/components/samsung_ac/protocol.cpp: In function 'uint16_t esphome::samsung_ac::skip_data(std::vector<unsigned char>&, int)':
src/esphome/components/samsung_ac/protocol.cpp:39:29: error: no matching function for call to 'find(__gnu_cxx::__normal_iterator<unsigned char*, std::vector<unsigned char> >, std::vector<unsigned char>::iterator, int)'
   39 |             return std::find(data.begin() + from, data.end(), 0x32) - data.begin();
      |                    ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/locale_facets.h:48,
                 from /data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_ios.h:37,
                 from /data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/ios:46,
                 from /data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/ostream:40,
                 from /data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/iostream:41,
                 from src/esphome/components/samsung_ac/util.h:4,
                 from src/esphome/components/samsung_ac/protocol.h:5,
                 from src/esphome/components/samsung_ac/protocol.cpp:1:
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/streambuf_iterator.h:435:5: note: candidate: 'template<class _CharT2> typename __gnu_cxx::__enable_if<std::__is_char<_CharT2>::__value, std::istreambuf_iterator<_CharT, std::char_traits<_CharT> > >::__type std::find(istreambuf_iterator<_CharT, char_traits<_CharT> >, istreambuf_iterator<_CharT, char_traits<_CharT> >, const _CharT2&)'
  435 |     find(istreambuf_iterator<_CharT> __first,
      |     ^~~~
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/streambuf_iterator.h:435:5: note:   template argument deduction/substitution failed:
src/esphome/components/samsung_ac/protocol.cpp:39:29: note:   '__gnu_cxx::__normal_iterator<unsigned char*, std::vector<unsigned char> >' is not derived from 'std::istreambuf_iterator<_CharT, std::char_traits<_CharT> >'
   39 |             return std::find(data.begin() + from, data.end(), 0x32) - data.begin();
      |                    ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*** [.pioenvs/tepelko/src/esphome/components/samsung_ac/protocol.cpp.o] Error 1
========================= [FAILED] Took 59.34 seconds =========================

--- a/components/samsung_ac/protocol.cpp
+++ b/components/samsung_ac/protocol.cpp
@@ -1,4 +1,5 @@
 #include "protocol.h"
+#include <algorithm>
 #include "util.h"
 #include "log.h"
 #include "protocol_nasa.h"
@@ -36,7 +37,10 @@ namespace esphome
             // and that one new proper packet will continue with the next data read
 
             // find next value of 0x32, and retry with that one
-            return std::find(data.begin() + from, data.end(), 0x32) - data.begin();
+            // return std::find(data.begin() + from, data.end(), 0x32) - data.begin();
+            return std::find(data.begin() + from, data.end(),
+                             static_cast<unsigned char>(0x32))
+                   - data.begin();
         }
 
         // This functions is designed to run after a new value was added

